Add docker dev prod workflow guide (#1094)
run-house/runhouse: docs/guides/docker-dev.rst

=================================================

.. raw:: html

    <p><a href="https://colab.research.google.com/github/run-house/notebooks/blob/stable/docs/guides/docker-dev.ipynb">
    <img height="20px" width="117px" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a></p>

This guide demonstrates how to use the same Docker image with your
Runhouse cluster, for both:

-  **Production**: running functions and code that is pre-installed on
   the Docker image
-  **Local development**: making local edits to your repo, and having
   local changes propagated over to the cluster for experimentation

Afterwards, we provide a script that shows how to easily set up and
toggle between these two settings, using the same cluster setup.

In this example, we are going to be using the `DJLServing 0.27.0 with
DeepSpeed
0.12.6 <https://github.com/aws/deep-learning-containers/blob/master/available_images.md#large-model-inference-containers>`__
Container, which includes HuggingFace Tranformers (4.39.0), Diffusers
(0.16.0), and Accelerate (0.28.0). We will use both the container
version of these packages, as well as local editable versions, of the

> diffusers and transformers packages to showcase both production ready

Reviewer: Delete "of the diffusers and transformers packages"

set some environment variables necessary to pull the Docker image.

For more specific details on getting your Docker image set up with
Runhouse, please take a look at <Guide: Docker Cluster Setup>.

.. code:: ipython3

    ! export SKYPILOT_DOCKER_USERNAME=AWS
    ! export SKYPILOT_DOCKER_PASSWORD=$(aws ecr get-login-password --region us-west-1)
    ! export SKYPILOT_DOCKER_SERVER=763104351884.dkr.ecr.us-west-1.amazonaws.com

Once these variables are set, we can import runhouse and construct an
ondemand cluster, specifying the container image id as follows, and call
``cluster.up_if_not()`` to launch the cluster with the Docker image
loaded on it.

.. code:: ipython3

    import runhouse as rh

.. parsed-literal::
    :class: code-output

    INFO | 2024-08-01 02:18:48.921683 | Loaded Runhouse config from /Users/caroline/.rh/config.yaml

.. code:: ipython3

    cluster = rh.ondemand_cluster(
        name="diffusers_docker",
        image_id="docker:djl-inference:0.27.0-deepspeed0.12.6-cu121",
        instance_type="g5.8xlarge",
        provider="aws",
    )
    cluster.up_if_not()

The function we’ll be using in our demo is ``is_transformers_available``

> from ``diffusers.utils``. We’ll first show what using this function like

Reviewer: ```suggestion
from ``diffusers.utils``. We’ll first show what using this function
```

directly on the box (e.g. a production setting) looks like. After, we’ll

> show the case if we had a local version of the repositories, that we
> modified locally, and wanted to test out the change on the cluster.

Reviewer: ```suggestion
show the case if we had local versions of the repositories that we'd
modified and wanted to test out our changes on the cluster.
```

.. code:: ipython3

    from diffusers.utils import is_transformers_available

Production Workflow
-------------------

> The core of the production workflow is that the Docker image consists of
> the exact packages and versions that you want. Regardless of whether or
> not one has local edits, these edits should **not** be used in the
> production workflow.

Reviewer: ```suggestion
The core of the production workflow is that the Docker image already contains
the exact packages and versions we want, probably published into the registry in CI/CD.
We don't want to perform any installs or code changes within the image throughout
execution so we preserve exact reprodicubility.
```

> **NOTE**: By default, Ray and Runhouse are installed on the ondemand
> cluster during setup time. This may be the most updated version of
> Runhouse, or the version you have cloned and installed locally. To make
> sure that your version is pinned for production use cases, please make
> sure that you have Runhouse and Ray included on the Docker container;
> the setup will skip the installations if it detects that they are
> already installed on the cluster.

Reviewer: ```suggestion
**NOTE**: By default, Ray and Runhouse are installed on the ondemand
cluster during setup time (generally attempting to match the versions you 
have locally), unless we detect that they're already present. To make
sure that no installs occur in production, please make
sure that you have Runhouse and Ray installed in your docker image.
```
